internal/catalogmetadata/cache/cache.go

```diff
@@ -117,34 +116,49 @@ func (fsc *filesystemCache) FetchCatalogContents(ctx context.Context, catalog *c
 	// the cached contents
 	if data, ok := fsc.cacheDataByCatalogName[catalog.Name]; ok {
 		if data.ResolvedRef == catalog.Status.ResolvedSource.Image.Ref {
-			return os.Open(cacheFilePath)
+			return os.DirFS(cacheDir), nil
 		}
 	}
 
-	if err = os.MkdirAll(cacheDir, os.ModePerm); err != nil {
-		return nil, fmt.Errorf("error creating cache directory for Catalog %q: %s", catalog.Name, err)
-	}
-
-	file, err := os.Create(cacheFilePath)
+	tmpDir, err := os.MkdirTemp(fsc.cachePath, fmt.Sprintf(".%s-", catalog.Name))
 	if err != nil {
-		return nil, fmt.Errorf("error creating cache file for Catalog %q: %s", catalog.Name, err)
+		return nil, fmt.Errorf("error creating temporary directory to unpack catalog metadata: %v", err)
 	}
 
-	if _, err := io.Copy(file, resp.Body); err != nil {
-		return nil, fmt.Errorf("error writing contents to cache file for Catalog %q: %s", catalog.Name, err)
+	if err := declcfg.WalkMetasReader(resp.Body, func(meta *declcfg.Meta, err error) error {
+		if err != nil {
+			return fmt.Errorf("error parsing catalog contents: %v", err)
+		}
+		pkgName := meta.Package
+		if meta.Schema == declcfg.SchemaPackage {
+			pkgName = meta.Name
+		}
+		metaName := meta.Name
+		if meta.Name == "" {
+			metaName = meta.Schema
+		}
+		metaPath := filepath.Join(tmpDir, pkgName, meta.Schema, metaName+".json")
+		if err := os.MkdirAll(filepath.Dir(metaPath), os.ModePerm); err != nil {
+			return fmt.Errorf("error creating directory for catalog metadata: %v", err)
+		}
+		if err := os.WriteFile(metaPath, meta.Blob, os.ModePerm); err != nil {
+			return fmt.Errorf("error writing catalog metadata to file: %v", err)
+		}
+		return nil
+	}); err != nil {
+		return nil, err
 	}
 
-	if err = file.Sync(); err != nil {
-		return nil, fmt.Errorf("error syncing contents to cache file for Catalog %q: %s", catalog.Name, err)
+	if err := os.RemoveAll(cacheDir); err != nil {
+		return nil, fmt.Errorf("error removing old cache directory: %v", err)
 	}
-
-	if _, err = file.Seek(0, 0); err != nil {
-		return nil, fmt.Errorf("error resetting offset for cache file reader for Catalog %q: %s", catalog.Name, err)
+	if err := os.Rename(tmpDir, cacheDir); err != nil {
+		return nil, fmt.Errorf("error moving temporary directory to cache directory: %v", err)
```

Comment:
curious to know what happens if fail here? Just run of the mill expbackoff? what happens to resolution during that period? Same thing?

Reply:
Yes to both.